envoy/upstream/outlier_detection.h

```diff
@@ -180,6 +275,31 @@ class EventLogger {
 
 using EventLoggerSharedPtr = std::shared_ptr<EventLogger>;
 
+using ExtMonitorCreateFn = std::function<ExtMonitorPtr()>;
+
+class ExtMonitorFactoryContext {
+public:
+  ExtMonitorFactoryContext(ProtobufMessage::ValidationVisitor& validation_visitor)
+      : validation_visitor_(validation_visitor) {}
+  ProtobufMessage::ValidationVisitor& messageValidationVisitor() const {
+    return validation_visitor_;
+  }
+
+private:
+  ProtobufMessage::ValidationVisitor& validation_visitor_;
+};
+
+class ExtMonitorFactory : public Envoy::Config::TypedFactory {
+public:
+  ~ExtMonitorFactory() override = default;
+
+  virtual ExtMonitorCreateFn createMonitor(const std::string& name,
+                                           ProtobufTypes::MessagePtr&& config,
+                                           const ExtMonitorFactoryContext& context) PURE;
```

Comment:
Hmmm, basically, we will use `const Protobuf::Message& config` as input parameter.

Reply:
cc @cpakulski WDYT?

Reply:
I am not sure what you mean here. `createMonitor` function returns a pointer to a function which creates a monitor of a specific type. That pointer is called later on, when actual monitor is created. The config must be somehow "captured" to be used later. So, the config is created here: https://github.com/envoyproxy/envoy/blob/81201ae640930b9e25a2e78f603c7fbb364955eb/source/common/upstream/outlier_detection_impl.cc#L316-L319 and "moved" to be used later on when a function is called. I may be missing something. Please let me know. Thx.

Reply:
Basically, in the factory method, we will use the `const Protobuf::Message&` as input config type. And in the factory method, the extension implementation need to load the proto config (and my do some validations to the proto config) to an extension-specific config object.
The loaded config object will be captured by the lambda and will be used when the lambda is called to create an acutal extension instance.

You can take the `createFilterFactoryFromProto` as an example.

Reply:
I am still missing something here.

> Basically, in the factory method, we will use the const Protobuf::Message& as input config type. And in the factory method, the extension implementation need to load the proto config (and my do some validations to the proto config) to an extension-specific config object.

So, the flow is that outlier config with typed extensions config is received here:
https://github.com/envoyproxy/envoy/pull/34154/files#diff-16c9f0f7c4c3df3cc89011911120d8c68b8c5249f11a80da5d9691aa4a060e57R308-R321

Factory is found here: https://github.com/envoyproxy/envoy/pull/34154/files#diff-16c9f0f7c4c3df3cc89011911120d8c68b8c5249f11a80da5d9691aa4a060e57R314

And config is translated here: https://github.com/envoyproxy/envoy/pull/34154/files#diff-16c9f0f7c4c3df3cc89011911120d8c68b8c5249f11a80da5d9691aa4a060e57R316-R317

The config after translation is validated here: https://github.com/envoyproxy/envoy/pull/34154/files#diff-5318169dc9b00303bee70136099bd8bd3a24210698328d9028316402c527836bR93-R99 and it is immediately captured in lambda. Captured validated config is used when actual monitor is created.

I think I missing something here. Is the problem with the flow or the types of parameters?

Reply:
I will add some more comments to the code. Hope I can explain it clearly.

Reply:
> and it is immediately captured in lambda. Captured validated config is used when actual monitor is created.

The validation here is only based on the PGV annotations. The specific monitor implementation may do more validations on required. We shouldn't capture the proto, but should capture a config object that loaded by the specific monitor implementation from the proto.

See https://github.com/envoyproxy/envoy/pull/34154#discussion_r1689039529 where I updated your code segment and add some comments to the code.